dns/route53/dns_test.go

```diff
@@ -0,0 +1,166 @@
+package route53
+
+import (
+	"testing"
+	"time"
+
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+
+const (
+	MOCKED_ZONE_ID = "20N31D"
+)
+
+
+type mockR53 struct {
+	recordSets map[string]*route53.ResourceRecordSet
+}
+
+func (r53 *mockR53) ChangeResourceRecordSets(input *route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error) {
+	for _, change := range input.ChangeBatch.Changes {
+		switch *change.Action {
+		case route53.ChangeActionCreate, route53.ChangeActionUpsert:
+			r53.recordSets[*change.ResourceRecordSet.Name] = change.ResourceRecordSet
+
+		case route53.ChangeActionDelete:
+			delete(r53.recordSets, *change.ResourceRecordSet.Name)
+
+		default:
+			panic(fmt.Sprintf("Invalid action: %s", *change.Action))
+		}
+	}
+
+	return &route53.ChangeResourceRecordSetsOutput{
+		ChangeInfo: &route53.ChangeInfo{
+			Id:          aws.String("MOCKED-AWS-API-ID"),
+			Status:      aws.String("INSYNC"),
+			SubmittedAt: aws.Time(time.Now()),
+		},
+	}, nil
+}
+
+func (r53 *mockR53) ListResourceRecordSets(input *route53.ListResourceRecordSetsInput) (*route53.ListResourceRecordSetsOutput, error) {
+	recordName := *input.StartRecordName
+
+	resourceRecordSets := []*route53.ResourceRecordSet{}
+	if r53.recordSets[recordName] != nil {
+		resourceRecordSets = append(resourceRecordSets, r53.recordSets[recordName])
+	}
+
+	return &route53.ListResourceRecordSetsOutput{
+		ResourceRecordSets: resourceRecordSets,
+		IsTruncated:        aws.Bool(false),
+		MaxItems:           input.MaxItems,
+	}, nil
+}
+
+// NOTE: always returns a list of one item matching the queried DNSName
+func (r53 *mockR53) ListHostedZonesByName(input *route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error) {
+	return &route53.ListHostedZonesByNameOutput{
+		HostedZones: []*route53.HostedZone{
+			&route53.HostedZone{
+				Id:              aws.String(MOCKED_ZONE_ID),
+				CallerReference: aws.String("MOCKED-CAL-REF"),
+				Name:            input.DNSName,
+			},
+		},
+		IsTruncated: aws.Bool(false),
+		MaxItems:    input.MaxItems,
+	}, nil
+}
+
+
+func TestAddAndRemove(t *testing.T) {
+	log.SetLevel(log.DebugLevel)
+
+	recordName := fmt.Sprintf("%s.%s", "_srv._tcp.test", "example.com")
+	ttl := uint16(60)
+	dryRun := false
+	client := &Client{
+		Service: &mockR53{
+			recordSets: map[string]*route53.ResourceRecordSet{},
+		},
+	}
+	hostedZoneID := MOCKED_ZONE_ID
+
+	srvManager := NewSRVManager(
+		client,
+		hostedZoneID,
+		recordName,
+		ttl,
+		dryRun,
+	)
+
+	record := &net.SRV{
+		Priority: 10,
+		Weight:   20,
+		Port:     4242,
+		Target:   "foobar.domain.tld",
```

Comment:
IIRC, `Target` is supposed to contain a dot at the end. (also applies to integration test below)

Reply:
If that is the case, then this needs to be

a) documented (maybe I'm just too blind to spot it, if that is already the case. I was able to only find a hint in the RFC's *Fictional example* section)
b) enforced somehow. `net.SRV` doc doesnt say anything about it.

Reply:
Again, with my RFC hat on, please keep example.com, or use anything else according to RFC2606.